File: 316-remove-duplicate-letters/remove-duplicate-letters.py
class Solution:
    def removeDuplicateLetters(self, s: str) -> str:
        freq={c: s.count(c) for c in set(s)}
        stack=[]
        in_stack=set()
        for c in s:
            freq[c]-=1
            if c in in_stack:
                continue
            while stack and c<stack[-1] and freq[stack[-1]]>0:
                removed=stack.pop()
                in_stack.remove(removed)
            stack.append(c)
            in_stack.add(c)
        return "".join(stack) 

File: 316-remove-duplicate-letters/remove-duplicate-letters.cpp
class Solution {
public:
  string removeDuplicateLetters(string s) {
    map<char, int> freq;
    for (char c : s) freq[c]++;

    map<char, bool> inResult;
    string result = "";

    for (char c : s) {
        freq[c]--;

        if (inResult[c]) continue;

        // pop from result if lexicographically larger
        while (!result.empty() && c < result.back() && freq[result.back()] > 0) {
            inResult[result.back()] = false;
            result.pop_back();
        }

        result.push_back(c);
        inResult[c] = true;
    }

    return result;
        
    }
};